pghoard/basebackup_delta.py

```diff
@@ -0,0 +1,347 @@
+        """Upload file into an object with a temporary name, after upload is finished, it will be copied to a new name
+        with a final/real hash"""
+        # length of UUID is 36 + prefix, should not conflict with actual digest file names
+        temp_object_name = f"temp_{uuid.uuid4()}"
+        self._delta_upload_hexdigest(
+            callback_queue=callback_queue,
+            chunk_path=f"{self.compressed_base}/{temp_object_name}",
+            temp_dir=self.temp_base_dir,
+            file_obj=file_obj,
+            relative_path=relative_path
+        )
+
+    def _list_existing_files(self):
+        """Iterate through all manifest files and fetch information about delta files"""
+        all_snapshot_files: Dict[str, SnapshotFile] = {}
+        for backup in self.get_remote_basebackups_info(self.site):
+            if backup["metadata"].get("format") != BaseBackupFormat.delta_v1:
+                continue
+
+            key = os.path.join(self.site_config["prefix"], "basebackup", backup["name"])
+            bmeta_compressed = self.storage.get_contents_to_string(key)[0]
+
+            with rohmufile.file_reader(
+                fileobj=io.BytesIO(bmeta_compressed),
+                metadata=backup["metadata"],
+                key_lookup=lambda key_id: self.site_config["encryption_keys"][key_id]["private"]
+            ) as input_obj:
+                meta = extract_pghoard_delta_v1_metadata(input_obj)
+
+            manifest = meta["manifest"]
+            snapshot_result = manifest["snapshot_result"]
+            backup_state = snapshot_result["state"]
+            files = backup_state["files"]
+            for delta_file in files:
+                snapshot_file = SnapshotFile.parse_obj(delta_file)
+                if snapshot_file.hexdigest:
+                    all_snapshot_files[snapshot_file.hexdigest] = snapshot_file
+
+        return all_snapshot_files
+
+    def _delta_upload_hexdigest(self, *, temp_dir, chunk_path, file_obj, callback_queue, relative_path):
+        start_time = time.monotonic()
+
+        input_size = file_obj.seek(0, os.SEEK_END)
+        file_obj.seek(0)
+
+        uploaded_hash = hashlib.blake2s()
+
+        def update_hash(data):
+            uploaded_hash.update(data)
+
+        with NamedTemporaryFile(dir=temp_dir, prefix=os.path.basename(chunk_path), suffix=".tmp") as raw_output_obj:
+            rohmufile.write_file(
+                input_obj=file_obj,
+                output_obj=raw_output_obj,
+                compression_algorithm=self.compression_data.algorithm,
+                compression_level=self.compression_data.level,
+                rsa_public_key=self.encryption_data.rsa_public_key,
+                log_func=self.log.info,
+                data_callback=update_hash
+            )
+            result_size = raw_output_obj.tell()
+            raw_output_obj.seek(0)
+
+            os.link(raw_output_obj.name, chunk_path)
+
```

Comment:
We could / should check the produced hash at this point and skip upload in case new hash matches that of any file that already exists remotely, also including files that we've processed in relation to this backup so that we don't unnecessarily upload multiple files all having the same hash.

Reply:
I'll add check for the current backup, but do you think it's worth to check remotely per each object? It cost money and slows down everything, also before the backup we do check which digests are already uploaded (from manifest files), so it's mainly relevant to the current backup.

Reply:
Un-resolve the conversation please if you still think I am missing out something (WRT current backup - added check)

Reply:
I didn't mean checking remotely, just from the list of hashes we've already obtained earlier from the delta metadata files.

Reply:
This is unlikely very relevant from the point-of-view of trying to optimize uploads but if proper cleanup is missing, as it looks to me, then only uploading files that weren't already there would make it easier to keep track of what we need to delete in cleanup.